enterprise/coderd/prebuilds/reconcile.go

```diff
@@ -0,0 +1,489 @@
+	if ctx.Err() != nil {
+		return nil, nil, ctx.Err()
+	}
+
+	state := snapshot.CalculateState()
+	actions, err := snapshot.CalculateActions(c.clock, c.cfg.ReconciliationBackoffInterval.Value())
+	return state, actions, err
+}
+
+func (c *StoreReconciler) Reconcile(ctx context.Context, ps prebuilds.PresetSnapshot, actions prebuilds.ReconciliationActions) error {
+	logger := c.logger.With(
+		slog.F("template_id", ps.Preset.TemplateID.String()),
+		slog.F("template_name", ps.Preset.TemplateName),
+		slog.F("template_version_id", ps.Preset.TemplateVersionID),
+		slog.F("template_version_name", ps.Preset.TemplateVersionName),
+		slog.F("preset_id", ps.Preset.ID),
+		slog.F("preset_name", ps.Preset.Name),
+	)
+
+	prebuildsCtx := dbauthz.AsPrebuildsOrchestrator(ctx)
+
+	levelFn := logger.Debug
+	switch actions.ActionType {
+	case prebuilds.ActionTypeBackoff:
+		levelFn = logger.Warn
+	case prebuilds.ActionTypeCreate, prebuilds.ActionTypeDelete:
+		// Log at info level when there's a change to be effected.
+		levelFn = logger.Info
+	}
+
+	fields := []any{
+		slog.F("action_type", actions.ActionType),
+		slog.F("create_count", actions.Create),
+		slog.F("delete_count", len(actions.DeleteIDs)),
+		slog.F("to_delete", actions.DeleteIDs),
+	}
+	levelFn(ctx, "reconciliation actions for preset are calculated", fields...)
+
+	// TODO: add quartz
+	// TODO: i've removed the surrounding tx, but if we restore it then we need to pass down the store to these funcs.
+
+	switch actions.ActionType {
+	case prebuilds.ActionTypeBackoff:
+		// If there is anything to backoff for (usually a cycle of failed prebuilds), then log and bail out.
+		levelFn(ctx, "template prebuild state retrieved, backing off",
+			append(fields,
+				slog.F("backoff_until", actions.BackoffUntil.Format(time.RFC3339)),
+				slog.F("backoff_secs", math.Round(actions.BackoffUntil.Sub(c.clock.Now()).Seconds())),
+			)...)
+
+		// return ErrBackoff
+		return nil
+
+	case prebuilds.ActionTypeCreate:
+		var multiErr multierror.Error
+
+		for range actions.Create {
+			if err := c.createPrebuild(prebuildsCtx, uuid.New(), ps.Preset.TemplateID, ps.Preset.ID); err != nil {
+				logger.Error(ctx, "failed to create prebuild", slog.Error(err))
+				multiErr.Errors = append(multiErr.Errors, err)
+			}
+		}
+
+		return multiErr.ErrorOrNil()
+
+	case prebuilds.ActionTypeDelete:
+		var multiErr multierror.Error
+
+		for _, id := range actions.DeleteIDs {
+			if err := c.deletePrebuild(prebuildsCtx, id, ps.Preset.TemplateID, ps.Preset.ID); err != nil {
+				logger.Error(ctx, "failed to delete prebuild", slog.Error(err))
+				multiErr.Errors = append(multiErr.Errors, err)
+			}
+		}
+
+		return multiErr.ErrorOrNil()
+
+	default:
+		return xerrors.Errorf("unknown action type: %s", actions.ActionType)
+	}
+}
+
+func (c *StoreReconciler) createPrebuild(ctx context.Context, prebuildID uuid.UUID, templateID uuid.UUID, presetID uuid.UUID) error {
```

Comment:
Take a look into createPrebuild, deletePrebuild, provision and reconcile_test.go.